src/VkLayer_GF_amber_scoop/src/buffer_copy.cc

```diff
@@ -0,0 +1,208 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "VkLayer_GF_amber_scoop/buffer_copy.h"
+
+#include <VkLayer_GF_amber_scoop/vulkan_util.h>
+
+#include "gf_layers_layer_util/logging.h"
+
+namespace gf_layers::amber_scoop_layer {
+
+BufferCopy::BufferCopy(
+    DeviceData* device_data, const VkBuffer& buffer, VkDeviceSize buffer_size,
+    VkQueue queue, VkCommandPool command_pool,
+    const std::vector<const CmdPipelineBarrier*>& pipeline_barriers)
+    : device_data_(device_data) {
+  VkDevice device = device_data_->device;
+
+  // Create a buffer where the data will be copied to.
+  {
+    VkBufferCreateInfo create_info = {};
+    create_info.sType = VK_STRUCTURE_TYPE_BUFFER_CREATE_INFO;
+    create_info.size = buffer_size;
+    create_info.usage = VK_BUFFER_USAGE_TRANSFER_DST_BIT;
+    create_info.sharingMode = VK_SHARING_MODE_EXCLUSIVE;
+
+    RequireSuccess(device_data_->vkCreateBuffer(device, &create_info, nullptr,
+                                                &buffer_copy_),
+                   "Failed to create buffer for the copied data.");
+  }
+
+  VkMemoryRequirements memory_requirements;
+  device_data_->vkGetBufferMemoryRequirements(device, buffer_copy_,
+                                              &memory_requirements);
+  {
+    VkMemoryAllocateInfo vkMemoryAllocateInfo = {};
+    vkMemoryAllocateInfo.sType = VK_STRUCTURE_TYPE_MEMORY_ALLOCATE_INFO;
+    vkMemoryAllocateInfo.allocationSize = memory_requirements.size;
+    vkMemoryAllocateInfo.memoryTypeIndex =
+        FindMemoryType(memory_requirements.memoryTypeBits,
+                       VK_MEMORY_PROPERTY_HOST_VISIBLE_BIT);
+    RequireSuccess(
+        device_data_->vkAllocateMemory(device, &vkMemoryAllocateInfo, nullptr,
+                                       &buffer_copy_memory_),
+        "Failed to allocate memory for buffer copy.");
+  }
+  RequireSuccess(device_data_->vkBindBufferMemory(device, buffer_copy_,
+                                                  buffer_copy_memory_, 0),
+                 "Failed binding memory for buffer copy.");
+
+  // Create command buffer
+  VkCommandBuffer command_buffer = nullptr;
+  VkCommandBufferAllocateInfo command_buffer_allocate_info = {};
+  command_buffer_allocate_info.sType =
+      VK_STRUCTURE_TYPE_COMMAND_BUFFER_ALLOCATE_INFO;
+  command_buffer_allocate_info.commandPool = command_pool;
+  command_buffer_allocate_info.level = VK_COMMAND_BUFFER_LEVEL_PRIMARY;
+  command_buffer_allocate_info.commandBufferCount = 1;
+  RequireSuccess(device_data_->vkAllocateCommandBuffers(
+                     device, &command_buffer_allocate_info, &command_buffer),
+                 "Failed to allocate command buffers.");
```

Comment:
If we create a dispatchable object, we must update the dispatch table pointer. This is a strange thing we have to do due to the way Vulkan layers work, I think. I will do a PR shortly to add the function.

Reply:
See: https://github.com/google/gf-layers/pull/15

Once rebased, add:

```
  // We have created a dispatchable object, so we must set the dispatch table
  // pointer.
  CopyDispatchTablePointer(device, command_buffer);
```